lib/Conversion/CGGIToTfheRust/CGGIToTfheRust.cpp

```diff
@@ -0,0 +1,278 @@
+#include "include/Conversion/CGGIToTfheRust/CGGIToTfheRust.h"
+
+#include "include/Dialect/CGGI/IR/CGGIDialect.h"
+#include "include/Dialect/CGGI/IR/CGGIOps.h"
+#include "include/Dialect/LWE/IR/LWEAttributes.h"
+#include "include/Dialect/LWE/IR/LWEDialect.h"
+#include "include/Dialect/LWE/IR/LWEOps.h"
+#include "include/Dialect/LWE/IR/LWETypes.h"
+#include "include/Dialect/TfheRust/IR/TfheRustDialect.h"
+#include "include/Dialect/TfheRust/IR/TfheRustOps.h"
+#include "include/Dialect/TfheRust/IR/TfheRustTypes.h"
+#include "lib/Conversion/Utils.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"           // from @llvm-project
+#include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"   // from @llvm-project
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/ImplicitLocOpBuilder.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+
+namespace mlir::heir {
+
+#define GEN_PASS_DEF_CGGITOTFHERUST
+#include "include/Conversion/CGGIToTfheRust/CGGIToTfheRust.h.inc"
+
+Type encrytpedUIntTypeFromWidth(MLIRContext *ctx, int width) {
+  // Only supporting unsigned types because the LWE dialect does not have a
+  // notion of signedness.
+  switch (width) {
+    case 2:
+      return tfhe_rust::EncryptedUInt2Type::get(ctx);
+    case 3:
+      return tfhe_rust::EncryptedUInt3Type::get(ctx);
+    case 4:
+      return tfhe_rust::EncryptedUInt4Type::get(ctx);
+    case 8:
+      return tfhe_rust::EncryptedUInt8Type::get(ctx);
+    case 10:
+      return tfhe_rust::EncryptedUInt10Type::get(ctx);
+    case 12:
+      return tfhe_rust::EncryptedUInt12Type::get(ctx);
+    case 14:
+      return tfhe_rust::EncryptedUInt14Type::get(ctx);
+    case 16:
+      return tfhe_rust::EncryptedUInt16Type::get(ctx);
+    case 32:
+      return tfhe_rust::EncryptedUInt32Type::get(ctx);
+    case 64:
+      return tfhe_rust::EncryptedUInt64Type::get(ctx);
+    case 128:
+      return tfhe_rust::EncryptedUInt128Type::get(ctx);
+    case 256:
+      return tfhe_rust::EncryptedUInt256Type::get(ctx);
+    default:
+      llvm_unreachable("Unsupported bitwidth");
+  }
+}
+
+// Seems like this would be better as a method on the
+// LWE_EncodingAttrWithScalingFactor class, but I still have the problem of the
+// type returned by getEncoding being a vanilla Attribute. Probably we need a
```

Comment:
I think this was because of the any class that allowed for any LWE encodings. 

This was really annoying honestly :/ Yeah, like you say, adding some c++ code for a common interface or code to handle variadic types might end up better

Reply:
I think the right solution is to use MLIR interfaces/traits, but I just haven't figured out how to make a custom interface yet.